### PR TITLE
Use TUF to download key/cert material

### LINKED ISSUE #25: Retrieve CTFE signing key via TUF
At the moment, we've decided to check in the CTFE public key and use it to verify SCTs. The way this should really work is that we should check in the root key (`root.json`) and use it to download the CTFE key via TUF.

### PR DESCRIPTION
**Pull request by jku**
Should fix #25 by using TUF to download all certs and keys. It is based on code by @joshuagl, @wxjdsr (co-authors on first commit in this branch) but it is a significant rewrite compared to their branches.

FYI also @woodruffw, @di,  @SantiagoTorres. 

I'm allowing edits by maintainers, go wild if you feel like it. 

Status (updated dec 22):
* Works for signing and verifying
* The embedded keys/certs are no longer used or included in git (except staging keys as test assets)
* if the default (production) or --staging flows are used, TUF is used to fetch key material
* TUF is also used in some other cases (code does the same thing as the previous implementation, and uses production key material if none is provided)
* the implementation peeks into python-tuf internals currently (to get access to top-level target metadata). This should be okay for now but a long-term solution might be  https://github.com/theupdateframework/python-tuf/issues/1995
* When the caches are hot (certs, keys and metadata have been downloaded already), every "sigstore verify" and "sigstore sign" (with the production or staging flows) now results in a minimum of two network  requests (404 and a small download). This [may be avoidable](https://github.com/theupdateframework/python-tuf/issues/2225) in future but currently always happens
* with empty caches we download ~10 files, in the order of 20kB altogether. These files are then cached for later use

Some TODO items (there are certainly more)
- testing: many existing tests now require network
  - [x] we could mock the requests for staging in the tests
  - [ ] longer term could make it possible to not do network requests when metadata is not expired
- [x] testing: the TUF module has no tests (it does get tested by all tests using staging though)
- [x] TUF metadata is stored in `~/.local/share/sigstore-python/tuf/<url>/` and targets are cached in `~/.cache/sigstore-python/tuf/<url>/ ` -- this is not great for windows
- [x] logging in python-tuf is pretty verbose at INFO: we'll have to change that at some level
- [x] error handling is almost non-existent, but let's handle in later PR #357 . The situations I've identified are:
  * python-tuf RepositoryError (metadata from repo is not valid in some way)
  * python-tuf DownloadError (trouble downloading things)
  * python-tuf OSError (trouble storing things on disk)
  * discovery errors in our _tuf module: the repository content was not what we expected  
- [x] should remove the embedded key/cert material from the repo

### CODE REVIEW

**Comment by jku** [comment]
Oh, a note on fulcio: I now feed all certs with usage "Fulcio" and status "Active" to Verifier with the assumption that
* these all form a chain
* cryptography copes with the (non-controlled) order of these certs

hopefully there are no problems with these assumptions

File: sigstore/_internal/rekor/client.py
Comment (jku) [inline sigstore/_internal/rekor/client.py:412]: This is a bug: should not use a DEFAULT_REKOR_URL here. Need to reconsider

**Comment by woodruffw** [comment]
> Oh, a note on fulcio: I now feed all certs with usage "Fulcio" and status "Active" to Verifier with the assumption that
> 
> * these all form a chain
> * cryptography copes with the (non-controlled) order of these certs
> 
> hopefully there are no problems with these assumptions

These assumptions should be fine! We throw everything into an OpenSSL X.509 store, and I don't believe it imposes any order of insertion.

File: sigstore/_internal/rekor/client.py
Comment (woodruffw) [reply sigstore/_internal/rekor/client.py:412]: Making sure I understand: the bug here is that this URL will be wrong with `with_updater` is used on `TrustUpdater.staging()`, correct?

File: sigstore/_internal/tuf.py
Comment (woodruffw) [inline sigstore/_internal/tuf.py:37]: NTS: Might be worth using <https://pypi.org/project/pyxdg/> or similar here.

I agree that it's not great for Windows; maybe there's an even higher wrapper that handles both XDG and Windows-preferred paths.

Comment (woodruffw) [reply sigstore/_internal/tuf.py:37]: (And macOS too, since `~/Library/...` is preferred over XDG-style paths.)

Comment (di) [reply sigstore/_internal/tuf.py:37]: I agree that https://pypi.org/project/appdirs/ is probably a good solution for this.

Comment (woodruffw) [reply sigstore/_internal/tuf.py:37]: SG! I'll do that in a bit.

File: sigstore/_internal/rekor/client.py
Comment (woodruffw) [reply sigstore/_internal/rekor/client.py:412]: This is causing `sigstore verify --staging` to fail, so I'll look in to it next.

Comment (woodruffw) [reply sigstore/_internal/rekor/client.py:412]: More generally, this suggests that we should really have a top-level configuration object, similar to a configurator, that unifies all of this state (since we should always use like-with-like, so there's no need for flexibility here).

Comment (woodruffw) [reply sigstore/_internal/rekor/client.py]: 3a8f026 has the initial workaround.

Comment (woodruffw) [reply sigstore/_internal/rekor/client.py]: Another option here is to let individual components manage their TUF state -- is there any risk in having `RekorClient.production()` create its own `TrustUpdater`, other than the performance hit with the current implementation?

Alternatively, maybe a `TrustUpdater` singleton.

**Comment by woodruffw** [comment]
Online tests now pass, but the forced offline tests are still failing (as expected).

**Comment by woodruffw** [comment]
> * [ ]  logging in python-tuf is pretty verbose at INFO: we'll have to change that at some level
> *

IMO this is worth an upstream PR -- looks like most of these should really be `DEBUG` level instead. I can look into that.

File: sigstore/_internal/rekor/client.py
Comment (jku) [reply sigstore/_internal/rekor/client.py]: Thanks for investigating.

Making TrustUpdater aware of the sigstore mechanisms like RekorClient seems wrong.... although there are edge cases like CTKeyring thatare more data types that could indeed be return values for TrustUpdater.

I think the original idea is still salvageable -- I just removed the RekorClient.production()/staging() constructors too greedily. Creating a single TrustUpdater and supplying that "down the call stack" should work (there are only a few places where it's needed so it's not too much work).
 
I will try modifying the RekorClient constructors and will ask you to review if it seems to work.

File: sigstore/_verify/verifier.py
Comment (jku) [inline sigstore/_verify/verifier.py:243]: just checking: is this change supposed to be here?

**Comment by jku** [comment]
out of the remaining issues, the lack of offline tests seems the most pressing. This happens because all uses of production or staging flows for sign() and verify() currently mean a minimum of two requests to  the TUF repository -- and tests use especially staging extensively.

There are potentially two separate avenues for solving this:
* in general (not just in the specific case of testing), sigstore-python would rather not make those requests if the local metadata is not expired: python-tuf does not support this as it would be against the specification... but I think adding that feature to python-tuf may still make sense. This is not going to be entirely trivial so I would not expect this to happen in the next week or two. I don't think working around python-tuf on this is a good idea.
* we could tweak the test environment so that the requests do not happen there: I have a few ideas which I plan to try out today

**Comment by jku** [comment]
There weren't actually as many offline tests as I expected... but I made the mock fixture anyway as it's not that big -- of course it does mean I end up adding the staging key material in test assets.

There are still some failing tests, I'm looking at those after lunch: at least it looks line the pytest.mark.parametrize happens even if the test is marked as "online" and fails if the parametrized setup requires a connection

**Comment by jku** [comment]
I think TrustUpdater should have a few tests:
* make sure the expected things are downloaded over network (based on initial cache contents): not more, not less. We can mock the network requests for staging. Testing this online for production might be useful but a little trickier (as we can't tell what should get downloaded as easily)
* make sure the `get_*()` return values are sensible: we could just hard  code the expected values for staging

**Comment by jku** [comment]
The only clear blocker right now is the exceptions review: I could use some help with that -- or I'm fine with leaving that to someone else. My understanding of the possible errors are:
* from python-tuf: RepositoryError (metadata from repo is not valid in some way: many subclasses of this error are possible but they all mean update is not possible)
* from: python-tuf DownloadError (trouble downloading things)
* from python-tuf: OSError (trouble storing things on disk)
* discovery errors in sigstore_internal.tuf module (repository content is not what we expected)

Any suggestions from maintainers?

File: test/unit/conftest.py
Comment (jku) [inline test/unit/conftest.py:157]: I suppose this should actually return a function that yields the content, otherwise the test leaves the file unclosed

File: sigstore/_verify/verifier.py
Comment (woodruffw) [reply sigstore/_verify/verifier.py:239]: Yep -- `RekorEntryMissing.signature` is meant to be a `str`, but we were populating it with the raw `bytes` from the materials by accident. I noticed that while (accidentally) trigging the error case by mixing up different Sigstore instances 🙂

**Comment by woodruffw** [comment]
> The only clear blocker right now is the exceptions review: I could use some help with that -- or I'm fine with leaving that to someone else. My understanding of the possible errors are:
> 
> * from python-tuf: RepositoryError (metadata from repo is not valid in some way: many subclasses of this error are possible but they all mean we cannot download what we need)
> * from: python-tuf DownloadError (trouble downloading things)
> * from python-tuf: OSError (trouble storing things on disk)
> * discovery and parsing errors in sigstore_internal.tuf module (repository content is not what we expected)
> 
> Any suggestions from maintainers?

I'll give this some thought, but I'm _mostly_ okay with us punting on this for the initial PR -- `sigstore-python` needs some exception refactoring anyways, and we could probably do a better job with the overall refactor by batching any handling here into those changes.

File: test/unit/test_sign.py
Comment (woodruffw) [inline test/unit/test_sign.py:36]: Is there a reason for removing this parametrization?

File: test/unit/conftest.py
Comment (woodruffw) [reply test/unit/conftest.py:157]: That's worth a `NOTE` or `TODO` in the source IMO, but not a big deal otherwise -- I'm okay with cleanup at process destruction time for unit tests 🙂

**Comment by woodruffw** [comment]
> There weren't actually as many offline tests as I expected... but I made the mock fixture anyway as it's not that big -- of course it does mean I end up adding the staging key material in test assets.

Thanks for doing this! IMO we should put `assets-staging-tuf/` under the pre-existing `assets/` dir, but otherwise this looks great.

File: test/unit/test_sign.py
Comment (jku) [reply test/unit/test_sign.py:36]: Yes, the parametrization setup happens even if --skip-online: and that setup currently does a network request... 

Now that I think about it, I guess the parameter could be the class, and we only instantiate inside the test. That might work?

Comment (woodruffw) [reply test/unit/test_sign.py:36]: That SGTM!

File: sigstore/_internal/tuf.py
Comment (woodruffw) [inline sigstore/_internal/tuf.py:118]: `bandit` flagged this -- is there a reason for this `assert` beyond maybe an error handling case that we should use an exception for?

**Comment by woodruffw** [comment]
This is looking pretty good to me! I think we can move it out of a draft for now; there are only a few small cleanup tasks that I think should get done before we merge.

File: sigstore/_cli.py
Comment (tetsuo-cpp) [inline sigstore/_cli.py:420]: If we end up doing this for a custom Fulcio URL, it'd be good to log a warning as this is probably not what the user intended.

This would partially address #324.

I can follow up with another PR if you feel this is out of scope.

File: sigstore/_internal/tuf.py
Comment (tetsuo-cpp) [inline sigstore/_internal/tuf.py:86]: We've been using f-strings for this type of thing elsewhere in the codebase.

Comment (woodruffw) [reply sigstore/_internal/tuf.py:86]: Yeah, let's normalize on f-strings here. @tetsuo-cpp feel free to push to this branch!

Comment (tetsuo-cpp) [reply sigstore/_internal/tuf.py]: [d9aa72c](https://github.com/sigstore/sigstore-python/pull/351/commits/d9aa72c2e04477bf5e1d0b818f9ac31d4c91a651)

**Comment by tetsuo-cpp** [comment]
I'm seeing some weird lint errors on this branch that I don't see in `main`, CI doesn't seem to have this issue as it's getting all the way to `bandit`: 
```
sigstore/_utils.py:161: error: Returning Any from function declared to return "bytes"  [no-any-return]
Found 1 error in 1 file (checked 24 source files)
```

Looks like that file hasn't even been touched.

**Comment by jku** [comment]
> I'm seeing some weird lint errors on this branch that I don't see in `main`, CI doesn't seem to have this issue as it's getting all the way to `bandit`:
> 
> ```
> sigstore/_utils.py:161: error: Returning Any from function declared to return "bytes"  [no-any-return]
> Found 1 error in 1 file (checked 24 source files)
> ```
> 
> Looks like that file hasn't even been touched.

This is unrelated, it's not showing up in main because William added an ignore in 0647d75.

File: sigstore/_cli.py
Comment (jku) [reply sigstore/_cli.py:420]: That sounds like a good idea but since there's even an issue already, let's leave it for a future PR

**Comment by tetsuo-cpp** [comment]
> This is unrelated, it's not showing up in main because William added an ignore in [0647d75](https://github.com/sigstore/sigstore-python/commit/0647d759257d6974855547984ffef198f31b9803).

I assumed that was necessary somehow because we switched to Ruff, but your branch doesn't have that commit. Anyway, if you don't mind, I'll merge in `main` and get this branch up to date.

File: sigstore/_cli.py
Comment (tetsuo-cpp) [reply sigstore/_cli.py:420]: Yep, sounds good.

File: sigstore/_internal/tuf.py
Comment (jku) [reply sigstore/_internal/tuf.py:118]: A nicer solution would be great, the issue is this:
* `_trusted_set.targets` is correctly annotated `Optional[Metadata[Targets]]`
* the Optional is there because before the metadata is refreshed, `_trusted_set.targets` is None
* but at this point when metadata has been successfully refreshed we know that the actual annotation is `Metadata[Targets]`
* unfortunately mypy cannot understand this: the assert is there to inform mypy

So if you have other clever solutions, be my guest.

**Comment by jku** [comment]
I pushed some tiny documentation tweaks but I would agree: this looks good to me apart from the final linting issue https://github.com/sigstore/sigstore-python/pull/351#discussion_r1054711928

File: sigstore/_internal/tuf.py
Comment (jku) [reply sigstore/_internal/tuf.py:118]: if no smarter solutions appear, I would suggest a "# nosec" to disable bandit here: this feels like an appropriate use of assert to me.

Comment (tetsuo-cpp) [reply sigstore/_internal/tuf.py:118]: Agreed, I think that's fine.

**Review by tetsuo-cpp — DISMISSED** [review]
LGTM provided that the linting is sorted out. Nice work.

**Comment by jku** [comment]
I added the "#nosec" to get tests passing: feel free to solve the issue in some other way if you want.

File: sigstore/_internal/tuf.py
Comment (woodruffw) [reply sigstore/_internal/tuf.py]: I think we can use `typing.cast` for this! I'll push that up.

Comment (woodruffw) [reply sigstore/_internal/tuf.py]: Ah, looks like `cast` won't quite work. But I'm going to replace this with a `# type: ignore[union-attr]`, since that silences the underlying MyPy problem rather than turning it into a Bandit problem that we then ignore 🙂

**Review by woodruffw — APPROVED** [review]
LGTM! :shipit:

File: sigstore/_internal/tuf.py
Comment (jku) [reply sigstore/_internal/tuf.py]: that'll work but I'll note it's not just mypy: it's static typing in general for that code block. So e.g. editor predictions become unavailable as well when assert is removed, this is unfortunate.

I think the ultimate fix is to avoid Optional in cases like this. I will look at the python-tuf code and see what could be done.

File: sigstore/_internal/tuf.py
Comment (asraa) [inline sigstore/_internal/tuf.py:123]: note that on the verification path, Expired/Inactive targets can still be used to verify against. It might not be worth filtering these out, but rather returning a List of bytes that has this metadata attached

File: sigstore/_internal/tuf.py
Comment (asraa) [inline sigstore/_internal/tuf.py:121]: I actually would caution against using the custom metadata and use the path filters "rekor*", "fulcio*", "ctfe*", instead as the filtering mechanism. While we'll keep some of the metadata, I think we are aiming to change them (for example, status will be deprecated in favor of explicit NotBefore/NotAfter ranges)

Comment (asraa) [reply sigstore/_internal/tuf.py:121]: That being said "usage" should match here and be stable

File: sigstore/_internal/tuf.py
Comment (jku) [reply sigstore/_internal/tuf.py:123]: thanks -- I might add an argument like `active_only: bool` to the getters instead but this is a good point, not sure how I missed it.

I'll file a new issue on these, but maybe tomorrow

File: sigstore/_internal/tuf.py
Comment (jku) [reply sigstore/_internal/tuf.py:121]: Yeah I knew the custom metadata was not a permanent solution but I'm not sure filtering by path is going to be either -- maybe makes sense to update once the target discovery and delegation plans are properly decided

Comment (asraa) [reply sigstore/_internal/tuf.py:121]: > I'm not sure filtering by path is going to be either -- maybe makes sense to update once the target discovery and delegation plans are properly decided

That's fair! Yeah, in the end distributing as a single bundle would probably fix that and that likely will be the case next year. The migration to that will be slow anyway (as in, the existing targets will continue to be distributed like this).

File: sigstore/_internal/tuf.py
Comment (jku) [reply sigstore/_internal/tuf.py:123]: following up in #375

**Pull request merged**
